FreeRTOS_Sockets.c

```diff
@@ -3482,8 +3479,9 @@ void vSocketWakeUpUser( FreeRTOS_Socket_t * pxSocket )
 
                     xBytesLeft -= xByteCount;
 
-                    if( xBytesLeft == 0 )
+                    if( ( xBytesLeft == 0 ) || ( pvBuffer == NULL ) )
```

Comment:
If this is to be only used with zero copy, why should we not enclose this in an `#if` block? Because, if we are not using zero copy method, then this should fail(?).

Reply:
We could have introduced a new flag, but I'm afraid it is too late now. Some developers have been using this feature for many years. If we add a macro `ipconfig_TCP_SEND_ALLOWS_ZERO_COPY` it should be **enabled** by default, in order not to break existing code.

Background: as you know, TCP sockets have 2 stream buffers: one for incoming and one for outgoing. This has advantages, but it makes it impossible to create a zero-copy feature like UDP sockets have. Zero-copy UDP sockets use network buffers, zero-copy TCP sockets use pointers to the stream buffers.

The zero-copy feature in TCP has always been classified as "for advanced users only".